src/integrations/terminal/TerminalRegistry.ts

```diff
@@ -1,22 +1,74 @@
 import * as vscode from "vscode"
-
-export interface TerminalInfo {
-	terminal: vscode.Terminal
-	busy: boolean
-	lastCommand: string
-	id: number
-	stream?: AsyncIterable<string>
-	running: boolean
-	streamClosed: boolean
-}
+import { arePathsEqual } from "../../utils/path"
+import { Terminal } from "./Terminal"
+import { TerminalProcess } from "./TerminalProcess"
 
 // Although vscode.window.terminals provides a list of all open terminals, there's no way to know whether they're busy or not (exitStatus does not provide useful information for most commands). In order to prevent creating too many terminals, we need to keep track of terminals through the life of the extension, as well as session specific terminals for the life of a task (to get latest unretrieved output).
 // Since we have promises keeping track of terminal processes, we get the added benefit of keep track of busy terminals even after a task is closed.
 export class TerminalRegistry {
-	private static terminals: TerminalInfo[] = []
+	private static terminals: Terminal[] = []
 	private static nextTerminalId = 1
+	private static disposables: vscode.Disposable[] = []
+	private static isInitialized = false
+
+	static initialize() {
+		if (this.isInitialized) {
+			throw new Error("TerminalRegistry.initialize() should only be called once")
+		}
+		this.isInitialized = true
+
+		try {
+			// onDidStartTerminalShellExecution
+			const startDisposable = vscode.window.onDidStartTerminalShellExecution?.(
+				async (e: vscode.TerminalShellExecutionStartEvent) => {
+					// Get a handle to the stream as early as possible:
+					const stream = e?.execution.read()
+					const terminalInfo = this.getTerminalByVSCETerminal(e.terminal)
+					if (terminalInfo) {
+						terminalInfo.setActiveStream(stream)
```

Comment:
If `stream` is `undefined` here, how are we preventing the task UI from freezing? It seems like we'll be waiting on `const stream = await streamAvailable` forever.

Reply:
Notice the detailed error in bold which is now applied to all various no_shell_integration issues for better feedback to the user and for our troubleshooting:

![image](https://github.com/user-attachments/assets/8b0d8e48-2998-40f2-a8d7-4e8b60499091)

and the link takes us to http://docs.roocode.com/troubleshooting/shell-integration/